syntax = "proto3";
import "google/protobuf/wrappers.proto";
package kvraft;

message Op{
    enum OpName{
        GET = 0;
        PUT = 1;
        DEL = 2;
    }
    string Key = 1;
    string Value = 2;
}

message LogEntry{
    Op Command = 1;
    int32 Term = 2;
}

message RequestVoteArgs{
    int32 Term = 1;
    int32 CandidateId = 2;
    int32 LastLogIndex = 3;
    int32 LastLogTerm = 4; 
}

message RequestVoteReply{
    int32 Term = 1;
    bool VoteGranted = 2;
}

message AppendEntriesArgs{
    int32 Term = 1;
    int32 LeaderId = 2;
    int32 PrevLogIndex = 3;
    int32 PrevLogTerm = 4;
    int32 LeaderCommit = 5;
    repeated LogEntry Entries = 6;
}

message AppendEntriesReply{
    int32 Term = 1;
    bool Success = 2;
    int32 ReplicatedIndex = 3;
}

service KVRaft{
    rpc RequestVote(RequestVoteArgs) returns (RequestVoteReply);
    rpc AppendEntries(AppendEntriesArgs) returns (AppendEntriesReply);
}